bake/bake.go

```diff
@@ -200,52 +201,63 @@ func ParseFiles(files []File, defaults map[string]string) (_ *Config, err error)
 	}()
 
 	var c Config
-	var fs []*hcl.File
+	var composeFiles []File
+	var hclFiles []*hcl.File
 	for _, f := range files {
-		cfg, isCompose, composeErr := ParseComposeFile(f.Data, f.Name)
-		if isCompose {
-			if composeErr != nil {
-				return nil, composeErr
+		switch filepath.Ext(strings.ToLower(f.Name)) {
+		case ".yml", ".yaml":
+			composeFiles = append(composeFiles, f)
+		case ".json":
+			hf, diags := hclparse.NewParser().ParseJSON(f.Data, f.Name)
+			if diags.HasErrors() {
+				return nil, diags
 			}
-			c = mergeConfig(c, *cfg)
-			c = dedupeConfig(c)
-		}
-		if !isCompose {
-			hf, isHCL, err := ParseHCLFile(f.Data, f.Name)
-			if isHCL {
-				if err != nil {
-					return nil, err
-				}
-				fs = append(fs, hf)
-			} else if composeErr != nil {
-				return nil, fmt.Errorf("failed to parse %s: parsing yaml: %v, parsing hcl: %w", f.Name, composeErr, err)
-			} else {
-				return nil, err
+			hclFiles = append(hclFiles, hf)
+		default:
+			hf, diags := hclparse.NewParser().ParseHCL(f.Data, f.Name)
+			if diags.HasErrors() {
+				return nil, diags
 			}
+			hclFiles = append(hclFiles, hf)
+		}
+	}
+
+	if len(composeFiles) > 0 {
+		cfg, cmperr := ParseComposeFiles(composeFiles)
+		if cmperr != nil {
+			return nil, errors.Wrap(cmperr, "failed to parse compose file")
 		}
+		c = mergeConfig(c, *cfg)
+		c = dedupeConfig(c)
 	}
 
-	if len(fs) > 0 {
-		if err := hclparser.Parse(hcl.MergeFiles(fs), hclparser.Opt{
+	if len(hclFiles) > 0 {
+		if err := hclparser.Parse(hcl.MergeFiles(hclFiles), hclparser.Opt{
 			LookupVar:     os.LookupEnv,
 			Vars:          defaults,
 			ValidateLabel: validateTargetName,
 		}, &c); err.HasErrors() {
 			return nil, err
 		}
 	}
+
 	return &c, nil
 }
 
 func dedupeConfig(c Config) Config {
 	c2 := c
+	c2.Groups = make([]*Group, 0, len(c2.Groups))
+	for _, g := range c.Groups {
+		g.Targets = dedupSlice(g.Targets)
+		c2.Groups = append(c2.Groups, g)
+	}
```

Comment:
As I read this, we need to make a non-destructive copy of `c`? If that's the case, we should deref and copy `g` before making changes to `g.Targets`.